### PR TITLE
Several changes to the codegen backend organization

### PR DESCRIPTION
**Pull request by bjorn3**
* Split functions from values in cg_ssa `BackendTypes`.
* Remove `is_const_integral` function from `ConstMethods`.
* Actually register the invalid monomorphization of intrinsic long diagnostic and remove the `diagnostics` method from `CodegenBackends`, as it was unused.
* Add cg_ssa and cg_utils provided methods to `default_provide`, so codegen backend don't have to do it themself.

### CODE REVIEW

**Comment by Centril** [comment]
cc @rkruppe @eddyb

File: src/librustc_codegen_ssa/traits/backend.rs [hunk 1]
```diff
@@ -14,6 +14,8 @@ use syntax_pos::symbol::InternedString;
 
 pub trait BackendTypes {
     type Value: CodegenObject;
+    type FuncId: CodegenObject;
```
Comment (eddyb) [inline src/librustc_codegen_ssa/traits/backend.rs]: IMO we should use a more general name, like `Function`.

File: src/librustc_codegen_llvm/common.rs [hunk 2]
```diff
@@ -86,6 +86,8 @@ impl Funclet<'ll> {
 
 impl BackendTypes for CodegenCx<'ll, 'tcx> {
     type Value = &'ll Value;
+    type FuncId = &'ll Value;
```
Comment (eddyb) [inline src/librustc_codegen_llvm/common.rs]: Maybe we should have a newtype of `Value` to force this to be correctly typed (LLVM has a class for functions which IIRC is a subclass of `GlobalValue`). But this is a low-priority cleanup.

File: src/librustc_codegen_ssa/traits/builder.rs [hunk 3]
```diff
@@ -34,7 +34,7 @@ pub trait BuilderMethods<'a, 'tcx>:
     + HasTargetSpec
 
 {
-    fn new_block<'b>(cx: &'a Self::CodegenCx, llfn: Self::Value, name: &'b str) -> Self;
+    fn new_block<'b>(cx: &'a Self::CodegenCx, llfn: Self::FuncId, name: &'b str) -> Self;
```
Comment (eddyb) [inline src/librustc_codegen_ssa/traits/builder.rs]: At some point we'll have to replace all of these `ll` prefixes in variable names.

File: src/librustc_codegen_ssa/traits/misc.rs [hunk 4]
```diff
@@ -11,14 +11,14 @@ pub trait MiscMethods<'tcx>: BackendTypes {
         &self,
     ) -> &RefCell<FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Self::Value>>;
     fn check_overflow(&self) -> bool;
-    fn instances(&self) -> &RefCell<FxHashMap<Instance<'tcx>, Self::Value>>;
+    fn instances(&self) -> &RefCell<FxHashMap<Instance<'tcx>, Self::FuncId>>;
```
Comment (eddyb) [inline src/librustc_codegen_ssa/traits/misc.rs]: I don't think this is correct, isn't this used for both statics and functions?

File: src/librustc_codegen_llvm/common.rs [hunk 5]
```diff
@@ -388,3 +386,9 @@ pub fn struct_in_context(
 fn hi_lo_to_u128(lo: u64, hi: u64) -> u128 {
     ((hi as u128) << 64) | (lo as u128)
 }
+
+fn is_const_integral(v: &'ll Value) -> bool {
+    unsafe {
+        llvm::LLVMIsAConstantInt(v).is_some()
```
Comment (eddyb) [inline src/librustc_codegen_llvm/common.rs]: Another place where stricter types could match LLVM's hierarchy (again, not necessarily relevant to this PR): `LLVMConstIntGetZExtValue` could take a `&ConstantInt` which `LLVMIsAConstantInt` would return an `Option` thereof, making `const_to_opt_uint(v)` be `LLVMIsAConstantInt(v).map(LLVMConstIntGetZExtValue)`.

File: src/librustc_interface/passes.rs [hunk 6]
```diff
@@ -789,10 +789,15 @@ pub fn default_provide(providers: &mut ty::query::Providers<'_>) {
     cstore::provide(providers);
     lint::provide(providers);
     rustc_lint::provide(providers);
+    rustc_codegen_utils::symbol_names::provide(providers);
+    rustc_codegen_ssa::back::symbol_export::provide(providers);
+    rustc_codegen_ssa::base::provide_both(providers);
```
Comment (eddyb) [inline src/librustc_interface/passes.rs]: There should be a top-level `provide`/`provide_both` per crate.

File: src/librustc_codegen_ssa/traits/misc.rs (hunk 4 above)
Comment (bjorn3) [reply src/librustc_codegen_ssa/traits/misc.rs]: You are right:

https://github.com/rust-lang/rust/blob/152527f443c51517bb867befa93809ce5b9b1cd1/src/librustc_codegen_llvm/mono_item.rs#L34

**Review by eddyb — APPROVED** [review]

**Pull request merged**